Docs: Add hetzner provider how-to for deploying NKP (#679)
nebari-dev/nebari-docs: docs/docs/how-tos/providers/hetzner.mdx

## Prerequisites

### Hetzner Cloud account

You'll need a Hetzner Cloud account. If you don't have one, [sign up at console.hetzner.cloud](https://console.hetzner.cloud/).

- **Location:** pick a datacenter where you want your cluster. Choose from one of the [availability zones](https://docs.hetzner.com/cloud/general/locations/).
- **Cost:** Hetzner Cloud has no free tier; servers, load balancers, and volumes all bill from the moment they're created.

### API token

`nic` authenticates to Hetzner using a project-scoped API token:

1. Open [console.hetzner.cloud](https://console.hetzner.cloud/), select your project, and go to **Security → API Tokens**.
2. Click **Generate API token**, give it a description (e.g., `nic-deploy`), and choose **Read & Write**.
3. Copy the token immediately — it is shown only once.

You'll add this token to your `.env` file. See the [Secrets and credentials](#secrets-and-credentials) section below for details.

### Install nic

Follow the [Install NIC](/docs/get-started/install) guide to download and install the `nic` CLI for your platform.

### GitOps repository

> See [GitOps repository](/docs/how-tos/deploy#gitops-repository) in the deploy lifecycle guide.

Reviewer: ```suggestion
See [GitOps repository](/docs/how-tos/prepare-to-deploy#gitops-repository) in the Prepare to deploy guide.
```

### Secrets and credentials

From inside your GitOps repo clone, download the [template](https://github.com/nebari-dev/nebari-infrastructure-core/blob/main/.env.example):

```bash
cd /path/to/your-gitops-repo
curl -o .env https://raw.githubusercontent.com/nebari-dev/nebari-infrastructure-core/main/.env.example
```

> Then uncomment and fill in the Hetzner token and GitOps tokens. For `.gitignore` setup and GitOps token configuration, see [Secrets and credentials](/docs/how-tos/deploy#secrets-and-credentials) in the deploy lifecycle guide.

Reviewer: ```suggestion
Then uncomment and fill in the Hetzner token and GitOps tokens. For `.gitignore` setup and GitOps token configuration, see [Secrets and credentials](/docs/how-tos/prepare-to-deploy#secrets-and-credentials) in the Prepare to deploy guide.
```

```

:::note
In later steps, `<config-file>` refers to this local copy (`hetzner-config.yaml`).
:::

At minimum, edit these fields:

```yaml
project_name: my-cluster          # lowercase alphanumeric, hyphens, underscores
domain: nebari.example.com        # a hostname you own

certificate:
  type: letsencrypt
  acme:
    email: you@example.com        # required for Let's Encrypt; renewal notices go here

git_repository:
  url: "https://github.com/<your-org>/<your-gitops-repo>.git"
  path: clusters/my-cluster       # subdirectory in the repo; conventionally matches project_name
  auth:
    token_env: GIT_TOKEN          # matches GIT_TOKEN set in .env

cluster:
  hetzner:
    kubernetes_version: "1.32"
    # REQUIRED: datacenter location.
    location: fsn1
    node_groups:
      # Exactly one node group must have master: true.
      # Count must be odd (1, 3, 5) when running HA embedded etcd.
      master:
        instance_type: cpx31      # 4 vCPU / 8 GB — good general-purpose starting point
        count: 1
        master: true
      workers:
        instance_type: cpx31
        count: 2

> ```

Reviewer 2: This config does not match the example one but that's not a problem. I think we need to update our example config in the repo (not here) to be more consistent with the other providers.

Author: Got it. Yeah this is based off of my tested hetzner deployment. I'll open an issue in nebari-infrastructure-core to address there

For more configuration options, see the [Hetzner provider reference](https://github.com/nebari-dev/nebari-infrastructure-core/blob/main/docs/design-doc/appendix/16-configuration-reference.md#hetzner-provider-configuration).
For the full schema (autoscaling, `persist_data`, Longhorn options), see the [NIC configuration reference](https://github.com/nebari-dev/nebari-infrastructure-core/blob/main/docs/design-doc/appendix/16-configuration-reference.md).

## Deploy

> Run the deploy commands as described in [Deploy](/docs/how-tos/deploy#deploy) in the deploy lifecycle guide. Allow 10–15 minutes for the first deployment — k3s cluster creation followed by ArgoCD syncing foundational services.

Reviewer: ```suggestion
Run the deploy commands as described in [Deploy and verify](/docs/how-tos/deploy#deploy-and-verify) in the deploy lifecycle guide. Allow 10–15 minutes for the first deployment — k3s cluster creation followed by ArgoCD syncing foundational services.

```

See [DNS](/docs/how-tos/deploy#dns) in the deploy lifecycle guide for DNS setup (Cloudflare or manual A record). For full DNS configuration options, see the [NIC DNS configuration reference](https://github.com/nebari-dev/nebari-infrastructure-core/blob/main/docs/design-doc/appendix/16-configuration-reference.md#dns-provider-configuration).

## Verify

`nic` writes the kubeconfig to `~/.cache/nic/hetzner-k3s/<project_name>/kubeconfig`. Point `kubectl` at it:

```bash
export KUBECONFIG=~/.cache/nic/hetzner-k3s/<project_name>/kubeconfig
```

> Then follow the [Verify](/docs/how-tos/deploy#verify) steps in the deploy lifecycle guide to check the cluster and ArgoCD applications.

Reviewer: ```suggestion
Then follow the [Deploy and verify](/docs/how-tos/deploy#deploy-and-verify) steps in the deploy lifecycle guide to check the cluster and ArgoCD applications.

```

## First sign-in

See [First sign-in](/docs/how-tos/deploy#first-sign-in) in the deploy lifecycle guide.

## Update an existing deployment

To change something about a running cluster (scale a node group, add a `gpu` group, change instance types, restrict network access), edit your config and re-run the deploy commands as described in [Update an existing deployment](/docs/how-tos/deploy#update-an-existing-deployment) in the deploy lifecycle guide.

:::warning
Changing `location` or `project_name` triggers destructive resource recreation. Treat these as one-way decisions.
:::

## Destroy

Run the destroy commands as described in [Destroy](/docs/how-tos/deploy#destroy) in the deploy lifecycle guide.

> `nic destroy` removes the k3s cluster (all servers), the Hetzner load balancer, and automatically cleans up orphaned Longhorn/CSI block volumes and any load balancers left behind by in-cluster controllers. This is more automatic than the AWS teardown — you rarely need to clean up manually.

Reviewer 2: Small nit. I would just get rid of this last sentence:

> This is more automatic than the AWS teardown — you rarely need to clean up manually

I don't think it's adding anything and on line 159 you're still telling the reader to confirm that no resources are lingering.

Author: I agree. I removed it. Thanks for flagging that!

---

> title: Deploy NKP on Hetzner Cloud

Reviewer: Can you change this to Hetzner instead? I am working on  #629 and this will allow matching with the rest of the pages in the placeholder.
```suggestion
title: Hetzner
```
